lib/npg_pipeline/function/s3_archiver.pm

```diff
@@ -0,0 +1,265 @@
+  my @extensions = qw{cram cram.md5 cram.crai
+                      seqchksum sha512primesums512.seqchksum
+                      F0x900.stats F0xB00.stats F0xF04_target.stats
+                      bcfstats};
+  push @expected_files,
+    map { $product->file_path($dir_path, ext => $_) } @extensions;
+
+  my $qc_path = $product->qc_out_path($self->archive_path());
+
+  my @qc_extensions = qw{verify_bam_id.json};
+  push @expected_files,
+    map { $product->file_path($qc_path, ext => $_) } @qc_extensions;
+
+  @expected_files = sort @expected_files;
+
+  return @expected_files;
+}
+
+=head2 create
+
+  Arg [1]    : None
+
+  Example    : my $defs = $obj->create
+  Description: Create per-product data file function definitions.
+
+  Returntype : ArrayRef[npg_pipeline::function::definition]
+
+=cut
+
+sub create {
+  my ($self) = @_;
+
+  my $id_run         = $self->id_run();
+  my $archive_config = $self->_read_config();
+
+  my @definitions;
+
+  my $i = 0;
+  foreach my $product (@{$self->products->{data_products}}) {
+    if ($product->is_tag_zero_product) {
+      $self->info('Skipping archiving for tag zero product ',
+                  $product->file_name_root);
+      next;
+    }
+    if ($product->lims->is_control) {
+      $self->info('Skipping archiving for control product ',
+                  $product->file_name_root);
+      next;
+    }
+
+    my $sample     = $product->lims->sample_supplier_name;
+    $sample or $self->logcroak('Failed to get a supplier sample name',
+                               'for product ', $product->file_name_root);
+
+    my $study_id   = $product->lims->study_id;
+    $study_id or $self->logcroak('Failed to get a study_id',
+                                 'for product ', $product->file_name_root);
+
+    if (exists $archive_config->{$study_id}) {
+      $self->info(sprintf q{S3 archiving %s in study %s to bucket URL %s},
+                  $product->file_name_root, $study_id,
+                  $archive_config->{$study_id});
+    } else {
+      $self->info(sprintf q{Skipping S3 archiving %s in study %s},
+                  $product->file_name_root, $study_id);
+      next;
+    }
+
+    my $job_name = sprintf q{%s_%d_%d}, $ARCHIVE_EXECUTABLE, $id_run, $i;
+    my $base_url = $archive_config->{$study_id};
+    $self->debug(sprintf q{Using base URL '%s' for study %s},
+                 $base_url, $study_id);
+
+    my @aws_args = qw{--cli-connect-timeout 300
+                      --acl bucket-owner-full-control};
+
+    foreach my $file_path ($self->expected_files($product)) {
+      my $filename   = basename($file_path);
+      my $file_url   = "$base_url/$sample/$filename";
+
+      my $command = sprintf q{%s s3 cp %s %s %s},
+        $ARCHIVE_EXECUTABLE, join(q{ }, @aws_args), $file_path, $file_url;
+
+      $self->info(sprintf q{S3 archiving %s in study %s to %s},
+                  $file_path, $study_id, $file_url);
+
+      push @definitions,
+      npg_pipeline::function::definition->new
```

Comment:
Is there something extra which needs to be added here when returning `@definitions` with > 1 member? The jobs generated have the npg_pipeline log wrapper around them (needed for our big analysis jobs as wr only captures the beginning and end of logs) but they will overwrite each other.

Reply:
'composition' attribute should be copied from the product object to the definition object. wr job identifies and LSF array indexes are derived from composition objects. And, @dkj is correct here, output log file names

Reply:
Done